### PR TITLE
fix: fix deny access to prevent a regression

### PR DESCRIPTION
**Pull request by micbar**
## Regression in ocis 7.0.0-rc.3

After updating an ocis 5.0.9 to ocis 7.0.0-rc.3, existing denials are breaking the web client.

deny-regression

## Related Issue

- Fixes <issue_link>

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by ScharfViktor** [comment]
- [ ] TODO check deny feature during the upgrade test between 5.0.0 and 7.0.0 
- [ ] put `FRONTEND_OCS_ENABLE_DENIALS` to deprecated envs

**Comment by mmattel** [comment]
If this gets merged, I will do the needed translation on transifex for the new strings.

Note, if TX needs to be added quickly befor final releasing, one needs to initiate a manualy sync, I will do the TX and another sync needs to be done. Then all is up to date.

**Review by 2403905 — APPROVED** [review]

**Comment by ScharfViktor** [comment]
tested: 
- creating/changing/deleting share with role `cannot access` - works correct
- user can see/change/delete `cannot access` share which was created in 5.0 version

remarks:
- 500 error when deleting `deny share` (was created in version 5.0) but looks like works -> user recipient can see resource
steps: 
- switch ocis version 5
- `admin` creates folder and shares folder to `Users` group
- `admin` shares folder to `einstein` with `denied` role
- upgrade ocis to version 7
- `admin` deletes `denied` role 
log: 
`{"level":"error","service":"sharing","pkg":"rgrpc","traceid":"56df2dfcf09102480861bcf5efb79e17","hostname":"","userID":"9955764e-d4f7-4191-96ea-7d559e8f7955","shareID":"b1b4a413-4987-45ed-8cee-986911d246cd:9955764e-d4f7-4191-96ea-7d559e8f7955:fa3ce1ec-9cd8-475e-80f9-053e86c1c836","error":"error: already exists: already exists","time":"2024-11-22T17:23:14+01:00","message":"persisting removed share failed"}`

@micbar today I saw that you had similar issue and fixed it. it was case with project space - it works

- no german translation
image

- user can see activity of the denied resource
- sse events in this case work correct but confuse users. If we deny access -> user gets notification that resource was shared but user doesn't see resource. If we delete `cannot access` - user see notification that share was deleted and user can see resource. 
- user can search by denied resource and cannot create resource with same name - existing bugs https://github.com/owncloud/web/issues/9029 https://github.com/owncloud/ocis/issues/6288

**Comment by micbar** [comment]
> @micbar today I saw that you had similar issue and fixed it. it was case with project space - it works

Correct. That was a general issue with sharing and has been fixed in reva https://github.com/cs3org/reva/pull/4968.

Needs a reva bump to arrive in ocis master.

File: services/graph/pkg/service/v0/base.go [hunk 1]
```diff
@@ -482,6 +482,10 @@ func (g BaseGraphService) cs3UserShareToPermission(ctx context.Context, share *c
 		perm.SetRoles([]string{role.GetId()})
 	} else {
 		actions := unifiedrole.CS3ResourcePermissionsToLibregraphActions(share.GetPermissions().GetPermissions())
+		// neither a role nor actions are set, that is a full denial
+		if len(actions) == 0 {
+			actions = []string{"access denied"}
```
Comment (micbar) [inline services/graph/pkg/service/v0/base.go:487]: @rhafer Not the "nicest" solution, I admit.

We have the contract with the clients that we 
- either set a value on `roles`
- or set a value on `@libre.graph.permissions.actions`

That means, if we disable the role "RoleDenied" and have existing shares, we need to return at least a hint to the user.

Fine for me, if you want to solve it differently.

File: services/graph/pkg/service/v0/base.go [hunk 2]
```diff
@@ -482,6 +482,10 @@ func (g BaseGraphService) cs3UserShareToPermission(ctx context.Context, share *c
 		perm.SetRoles([]string{role.GetId()})
 	} else {
 		actions := unifiedrole.CS3ResourcePermissionsToLibregraphActions(share.GetPermissions().GetPermissions())
+		// neither a role nor actions are set, that is a full denial
```
Comment (micbar) [inline services/graph/pkg/service/v0/base.go:485]: Important Info: We decided long ago, that we will only go with a "full denial", we do not intend to think about negative permissions in the future. So my assumption here is valid.

File: services/graph/pkg/unifiedrole/roles.go [hunk 3]
```diff
@@ -439,6 +448,26 @@ var (
...
+			Id:          proto.String(UnifiedRoleDeniedID),
+			Description: proto.String(_deniedUnifiedRoleDescription),
+			DisplayName: proto.String(cs3RoleToDisplayName(r)),
+			RolePermissions: []libregraph.UnifiedRolePermission{
+				{
+					AllowedResourceActions: CS3ResourcePermissionsToLibregraphActions(r.CS3ResourcePermissions()),
+					Condition:              proto.String(UnifiedRoleConditionFile),
+				},
+				{
+					AllowedResourceActions: CS3ResourcePermissionsToLibregraphActions(r.CS3ResourcePermissions()),
```
Comment (micbar) [inline services/graph/pkg/unifiedrole/roles.go:464]: @hodyroff Didn't we decide to allow denials only on folders? Technically it makes not a big difference but from a user POV it feels weird to set that on single files.

Comment (micbar) [reply services/graph/pkg/unifiedrole/roles.go:464]: After discussion, decided to keep that only on folders. The struggle to not finding single denied files for the receiver and communicating this to the share owner could create a bad UX.

**Comment by mmattel** [comment]
> * put `FRONTEND_OCS_ENABLE_DENIALS` to deprecated envs

**Before merging**, if the var has been deprecated, please run `make -C docs docs-generate` to add the changed `env_vars.yaml` to the PR. I need to do afterwards with a separate PR the changed envvar magic... 😅

**Comment by micbar** [comment]
> Before merging, if the var has been deprecated, please run make -C docs docs-generate to add the changed env_vars.yaml to the PR. I need to do afterwards with a separate PR the changed envvar magic... 😅

Needs to be discussed with @kobergj 

We deprecated the whole OCS Api with 5.0.0.

I would have removed it already but the clients were not able to move to the LibreGraph API in time.

**Comment by butonic** [comment]
> We deprecated the whole OCS Api with 5.0.0.
7.0.0?

**Comment by micbar** [comment]
> > [We deprecated the whole OCS Api with 5.0.0.](https://github.com/owncloud/ocis/pull/10627#issuecomment-2494394490)
> 
> 7.0.0?

https://doc.owncloud.com/ocis_release_notes.html#deprecation-announcements

File: services/graph/pkg/service/v0/base.go (hunk 1 above)
Comment (micbar) [reply services/graph/pkg/service/v0/base.go]: After discussion with @butonic and @kobergj we decided to return `none` in that case.

File: services/graph/pkg/unifiedrole/roles.go [hunk 4]
```diff
@@ -161,6 +163,12 @@ var (
 	// UnifiedRole SecureViewer, Role DisplayName (resolves directly)
 	_secureViewerUnifiedRoleDisplayName = l10n.Template("Can view (secure)")
 
+	// UnifiedRole FullDenial, Role Description (resolves directly)
+	_deniedUnifiedRoleDescription = l10n.Template("Deny all access.")
+
+	// UnifiedRole FullDenial, Role DisplayName (resolves directly)
+	_deniedUnifiedRoleDisplayName = l10n.Template("Cannot access.")
```
[the remark this answers is not in the document]
Comment (micbar) [reply services/graph/pkg/unifiedrole/roles.go]: done.

**Review by kobergj — APPROVED** [review]

**Pull request merged**